src/SFA.DAS.Reservations.Application/BulkUpload/Queries/BulkValidateCommandHandler.cs

```diff
@@ -0,0 +1,220 @@
+    {
+        private readonly IAccountReservationService _accountReservationService;
+        private readonly IGlobalRulesService _globalRulesService;
+        private readonly IAccountLegalEntitiesService _accountLegalEntitiesService;
+        private readonly IAccountsService _accountService;
+        private readonly IMediator _mediator;
+        private readonly Dictionary<long, AccountLegalEntity> _cachedAccountLegalEntities;
+
+        public BulkValidateCommandHandler(IAccountReservationService accountReservationService,
+            IGlobalRulesService globalRulesService,
+            IAccountLegalEntitiesService accountLegalEntitiesService, IAccountsService accountsService
+            , IMediator mediator)
+        {
+            _accountReservationService = accountReservationService;
+            _globalRulesService = globalRulesService;
+            _accountLegalEntitiesService = accountLegalEntitiesService;
+            _accountService = accountsService;
+            _mediator = mediator;
+            _cachedAccountLegalEntities = new Dictionary<long, AccountLegalEntity>();
+        }
+
+        public async Task<BulkValidationResults> Handle(BulkValidateCommand bulkRequest, CancellationToken cancellationToken)
+        {
+            var result = new BulkValidationResults();
+            result.ValidationErrors = new List<BulkValidation>();
+
+            // Only run validation for valid agreement ids - which have values.
+            var validAgreementIds = bulkRequest.Requests.Where(x => x.AccountLegalEntityId.HasValue);
+
+            var groups = validAgreementIds.GroupBy(x => x.AccountLegalEntityId.Value);
+
+            foreach (var group in groups)
+            {
+                AccountLegalEntity accountLegalEntity = await GetAccountLegalEntity(group.Key);
+                if (accountLegalEntity != null)
+                {
+                    if (accountLegalEntity.AgreementSigned && !accountLegalEntity.IsLevy)
+                    {
+                        if (await ApprenticeshipCountExceedsRemainingReservations(accountLegalEntity.AccountId, group.Count()))
+                        {
+                            result.ValidationErrors.Add(new BulkValidation { Reason = $"The employer has reached their reservations limit. Contact the employer.", RowNumber = group.First().RowNumber });
+                        }
+                        else if (await FailedGlobalRuleValidation())
+                        {
+                            result.ValidationErrors.Add(new BulkValidation { Reason = "Failed global rule validation", RowNumber = group.First().RowNumber });
+                            return result;
+                        }
+                        else if (await FailedAccountRuleValidation(accountLegalEntity.AccountId))
+                        {
+                            result.ValidationErrors.Add(new BulkValidation { Reason = "Reservation Paused", RowNumber = group.First().RowNumber });
```

Comment:
is this the right error for this validation rule being tripped?

Reply:
As discussed, these ACs are not covered in the user story - they will need to be covered in future user stories.  Just adding them in as don't want miss them.